src/v/archival/archiver_operations_api.h

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "archival/async_data_uploader.h"
+#include "base/outcome.h"
+#include "cloud_storage/remote_segment_index.h"
+#include "cloud_storage/types.h"
+#include "model/fundamental.h"
+#include "model/ktp.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
+namespace ss = seastar;
+
+namespace archival {
+
+struct manifest_upload_result {
+    model::ntp ntp;
+    size_t num_put_requests{0};
+    size_t size_bytes{0};
+
+    bool operator==(const manifest_upload_result& other) const = default;
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const manifest_upload_result& arg);
+};
+
+struct reconciled_upload_candidate {
+    /// Partition
+    model::ntp ntp;
+    /// Stream of log data in cloud-storage format
+    ss::input_stream<char> payload;
+    /// Size of the data stream
+    size_t size_bytes;
+    /// Segment metadata
+    cloud_storage::segment_meta metadata;
+    // Transactional metadata
+    fragmented_vector<model::tx_range> tx;
+
+    reconciled_upload_candidate() = default;
+
+    reconciled_upload_candidate(
+      model::ntp ntp,
+      ss::input_stream<char> payload,
+      size_t size_bytes,
+      cloud_storage::segment_meta metadata,
+      fragmented_vector<model::tx_range> tx)
+      : ntp(std::move(ntp))
+      , payload(std::move(payload))
+      , size_bytes(size_bytes)
+      , metadata(metadata)
+      , tx(std::move(tx)) {}
+
+    reconciled_upload_candidate(const reconciled_upload_candidate&) = delete;
+    reconciled_upload_candidate(reconciled_upload_candidate&&) noexcept
+      = default;
+    reconciled_upload_candidate& operator=(const reconciled_upload_candidate&)
+      = delete;
+    reconciled_upload_candidate&
+    operator=(reconciled_upload_candidate&&) noexcept
+      = default;
+
+    // NOTE: the operator is needed for tests only, it doesn't check
+    // equality of byte streams.
+    bool operator==(const reconciled_upload_candidate& o) const noexcept;
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const reconciled_upload_candidate& s);
+};
+
+struct upload_candidate_search_parameters {
+    static constexpr auto default_initialized
+      = std::numeric_limits<size_t>::max();
+    model::ntp ntp;
+    // Current term of the archiver
+    model::term_id archiver_term;
```

Comment:
Why is this important as a search parameter? In my mind a search parameter should only focus on what it will take to build input streams from local storage -- is this not how this is used?